lib/ams_lazy_relationships/core/lazy_dig_method.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+module AmsLazyRelationships::Core
+  module LazyDigMethod
+    def lazy_dig(*relation_names)
+      relationships = {
+        multiple: false,
+        data: [{
+          serializer: self.class,
+          object: object
+        }]
+      }
+
+      relation_names.each do |relation_name|
+        lazy_dig_relationship!(relation_name, relationships)
+      end
+
+      objects = relationships[:data].map { |r| r[:object] }
+
+      relationships[:multiple] ? objects : objects.first
+    end
+
+    private
+
+    def lazy_dig_relationship!(relation_name, relationships)
+      relationships[:data].map! do |serializer:, object:|
+        next_objects = lazy_dig_next_objects!(relation_name, serializer, object)
+        next unless next_objects
+
+        relationships[:multiple] ||= next_objects.respond_to?(:to_ary)
```

Comment:
well, we always initialize this flag by false at the beginning

```ruby
    def lazy_dig(*relation_names)
      relationships = {
        multiple: false,
        # ...
      }
```

The idea is to maintain the singular/plural nature of  `lazy_dig` returned value based on singular/plural nature of the `chain` of requested relation names. I.e.

```ruby
lazy_dig(:company, :logo) # returns single AR record or nil
lazy_dig(:companies, :logo) # returns the array of AR records
lazy_dig(:company, :logos) # returns the array of AR records
```

In general, `lazy_dig` will return single AR object (or nil) if `every` relationship in a chain was a kind of singular association, and will return an array otherwise. So highlighted expression is hunting for the first plural relationship

```
relationships[:multiple] ||= next_objects.respond_to?(:to_ary)
```

. Then we decide what kind of return value will be as follows

```ruby
      relationships[:multiple] ? objects : objects.first
```